pygradle-plugin/src/main/groovy/com/linkedin/gradle/python/tasks/action/CreateVirtualEnvAction.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2016 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.gradle.python.tasks.action;
+
+import com.linkedin.gradle.python.extension.PythonDetails;
+import com.linkedin.gradle.python.tasks.execution.TeeOutputContainer;
+import com.linkedin.gradle.python.util.pip.PipConfFile;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.process.ExecResult;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.function.Consumer;
+
+public class CreateVirtualEnvAction {
+
+    private static Logger log = Logging.getLogger(CreateVirtualEnvAction.class);
+    private final Project project;
+    private final PythonDetails pythonDetails;
+    private final TeeOutputContainer container = new TeeOutputContainer();
+
+    public CreateVirtualEnvAction(Project project, PythonDetails pythonDetails) {
+        this.project = project;
+        this.pythonDetails = pythonDetails;
+    }
+
+    public void buildVenv(@Nullable Consumer<File> customize) {
+        PipConfFile pipConfFile = new PipConfFile(project, pythonDetails);
+
+        File packageDir = makeTempDir().toFile();
+        getPyGradleBootstrap(project).getFiles().forEach(file -> {
+            project.copy(copySpec -> {
+                copySpec.from(project.tarTree(file.getPath()));
+                copySpec.into(packageDir);
+                copySpec.eachFile(it -> {
+                    //Remove the virtualenv-<version> from the file.
```

Comment:
The API works in terms of `FileCopyDetails` not File or Path. So it would be a lot of converting, and we would end up with more complicated logic than we have now I think

Reply:
As far as I can see this is splitting parent directory.
The existing Java APIs are cleaner way of doing that than the indexing.
We can talk about it if I am missing something.

Reply:
The Java api's are good at getting parents, but they aren't good at getting the first directory in a string. The API doesn't give us `File`s or `Path`s, we get strings (docs: https://docs.gradle.org/current/javadoc/org/gradle/api/file/FileCopyDetails.html)  So I think sub-string splitting is the easiest way to do this without having to iteratively search up until we get to the root.

Reply:
Using `subpath` and `getNameCount` should do the job.

Reply:
Fine. I don't think it's any easier to understand, and means you need to understand some finer points of Path API. But sure.